iceoryx_platform/generic/source/logging.cpp

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) 2024 by Mathias Kraus <elboberido@m-hias.de>. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iceoryx_platform/logging.hpp"
+
+#include <atomic>
+#include <iostream>
+#include <mutex>
+#include <sstream>
+#include <thread>
+
+namespace
+{
+// NOLINTJUSTIFICATION only used in this file; should be fine
+// NOLINTNEXTLINE(readability-function-size)
+void default_log_backend(
+    const char* file, int line, const char* function, IceoryxPlatformLogLevel log_level, const char* msg)
+{
+    if (log_level == IceoryxPlatformLogLevel::IOX_PLATFORM_LOG_LEVEL_OFF)
+    {
+        return;
+    }
+
+    std::stringstream stream;
+    stream << file << ":" << line << " (" << function << ") ";
+
+    switch (log_level)
+    {
+    case IceoryxPlatformLogLevel::IOX_PLATFORM_LOG_LEVEL_FATAL:
+        stream << "[Fatal]";
+        break;
+    case IceoryxPlatformLogLevel::IOX_PLATFORM_LOG_LEVEL_ERROR:
+        stream << "[Error]";
+        break;
+    case IceoryxPlatformLogLevel::IOX_PLATFORM_LOG_LEVEL_WARN:
+        stream << "[Warn ]";
+        break;
+    case IceoryxPlatformLogLevel::IOX_PLATFORM_LOG_LEVEL_INFO:
+        stream << "[Info ]";
+        break;
+    case IceoryxPlatformLogLevel::IOX_PLATFORM_LOG_LEVEL_DEBUG:
+        stream << "[Debug]";
+        break;
+    case IceoryxPlatformLogLevel::IOX_PLATFORM_LOG_LEVEL_TRACE:
+        stream << "[Trace]";
+        break;
+    default:
+        stream << "[UNDEF]";
+        break;
+    }
+
+    stream << " " << msg;
+
+    // NOLINTJUSTIFICATION We want to flush the line with each log output; if performance matters a custom log backend can be used, e.g. the logger in hoofs
+    // NOLINTNEXTLINE(performance-avoid-endl)
+    std::cout << stream.str() << std::endl;
+}
+
+enum class LoggerExchangeState : uint8_t
+{
+    DEFAULT,
+    PENDING,
+    CUSTOM,
+};
+
+struct IceoryxPlatformLogger
+{
+    std::atomic<IceoryxPlatformLogBackend> log_backend{&default_log_backend};
+    std::atomic<LoggerExchangeState> logger_exchange_state{LoggerExchangeState::DEFAULT};
+};
+
+IceoryxPlatformLogger& active_logger(IceoryxPlatformLogBackend new_log_backend)
```

Comment:
I assume this code was copied from some place else?!

Reply:
Well, I got inspired :)

I couldn't use the approach from the hoofs logger since it was not possible to expose the atomics so I needed to get creative and came up with this solution. I might even move it to the hoofs logger since it does not require a mutex.